src/rust/engine/fs/fs_util/src/main.rs

```diff
@@ -0,0 +1,180 @@
+      store_dir,
+      e
+    )
+  })?;
+
+  match top_match.subcommand() {
+    ("file", Some(sub_match)) => {
+      match sub_match.subcommand() {
+        ("cat", Some(args)) => {
+          let fingerprint = Fingerprint::from_hex_string(args.value_of("fingerprint").unwrap())?;
+          match store.load_bytes(&fingerprint)? {
+            Some(bytes) => {
+              io::stdout().write(&bytes).unwrap();
+              Ok(())
+            }
+            None => Err(format!("File with fingerprint {} not found", fingerprint)),
+          }
+        }
+        ("save", Some(args)) => {
+          save_file(&store, &PathBuf::from(args.value_of("path").unwrap()))
+            .map(|(fingerprint, _)| println!("{}", fingerprint))
+        }
+        (_, _) => unimplemented!(),
+      }
+    }
+    ("directory", Some(sub_match)) => {
+      match sub_match.subcommand() {
+        ("save", Some(args)) => {
+          save_directory(&store, &PathBuf::from(args.value_of("source").unwrap()))
+            .map(|(fingerprint, _)| println!("{}", fingerprint))
+        }
+        (_, _) => unimplemented!(),
+      }
+    }
+    (_, _) => unimplemented!(),
+  }
+}
+
+fn save_file(store: &Store, path: &Path) -> Result<(Fingerprint, usize), String> {
+  let mut buf = Vec::new();
+  File::open(path)
+    .and_then(|mut f| f.read_to_end(&mut buf))
+    .map_err(|e| {
+      format!("Error reading file {:?}: {}", path, e.description())
+    })?;
+  Ok((store.store_file_bytes(&buf)?, buf.len()))
+}
+
+fn save_directory(store: &Store, root: &Path) -> Result<(Fingerprint, usize), String> {
+  let mut directory = bazel_protos::remote_execution::Directory::new();
+  for entry in fs::PosixFS::scandir_sync(fs::Dir(root.to_path_buf()), &root)
+    .map_err(|e| {
+      format!("Error listing directory {:?}: {}", root, e.description())
+    })?
+  {
+    match entry {
+      fs::Stat::Dir(fs::Dir(path)) => {
+        let (fingerprint, proto_size_bytes) = save_directory(store, &path)?;
+        directory.mut_directories().push({
+          let mut dir_node = bazel_protos::remote_execution::DirectoryNode::new();
+          dir_node.set_name(file_name_as_utf8(&path)?);
+          dir_node.set_digest(fingerprint_to_digest(&fingerprint, proto_size_bytes as i64));
+          dir_node
+        })
+      }
+      fs::Stat::File(fs::File(path)) => {
+        let (fingerprint, size_bytes) = save_file(store, &path)?;
+        directory.mut_files().push({
+          let mut file_node = bazel_protos::remote_execution::FileNode::new();
+          file_node.set_name(file_name_as_utf8(&path)?);
+          file_node.set_digest(fingerprint_to_digest(&fingerprint, size_bytes as i64));
+          let mut absolute_path = root.to_path_buf();
+          absolute_path.push(path.file_name().ok_or_else(|| {
+            format!("{:?} did not have a file_name", path)
+          })?);
+          file_node.set_is_executable(
+            std::fs::metadata(&absolute_path)
```

Comment:
Maybe consider adding the execute bit to `Stat::File` in order to get it during the scandir?

Reply:
Yeah, I was thinking about that... I guess it's probably worthwhile :) Will do as a follow-up.